testing/fil_token_integration/actors/test_actor/src/lib.rs

```diff
@@ -0,0 +1,184 @@
+/// eg. transfer and then the receiver hook rejects:
+/// action(TestAction::Transfer(
+///         some_address,
+///         action(TestAction::Reject),
+///     ),
+/// )
+pub fn action(action: TestAction) -> RawBytes {
+    RawBytes::serialize(action).unwrap()
+}
+
+#[no_mangle]
+fn invoke(input: u32) -> u32 {
+    std::panic::set_hook(Box::new(|info| {
+        sdk::vm::abort(ExitCode::USR_ASSERTION_FAILED.value(), Some(&format!("{}", info)))
+    }));
+
+    let method_num = sdk::message::method_number();
+    match_method!(method_num, {
+        "Constructor" => {
+            NO_DATA_BLOCK_ID
+        },
+        "Receive" => {
+            // Received is passed a UniversalReceiverParams
+            let params: UniversalReceiverParams = deserialize_params(input);
+
+            // reject if not an FRC46 token
+            // we don't know how to inspect other payloads here
+            if params.type_ != FRC46_TOKEN_TYPE {
+                panic!("invalid token type, rejecting transfer");
+            }
+
+            // get token transfer data
+            let token_params: FRC46TokenReceived = params.payload.deserialize().unwrap();
+
+            // todo: examine the operator_data to determine our next move
+            let action: TestAction = token_params.operator_data.deserialize().unwrap();
+            match action {
+                TestAction::Accept => {
+                    // do nothing, return success
+                    NO_DATA_BLOCK_ID
+                }
+                TestAction::Reject => {
+                    // abort to reject transfer
+                    sdk::vm::abort(
+                        ExitCode::USR_FORBIDDEN.value(),
+                        Some("rejecting transfer"),
+                    );
+                }
+                TestAction::Transfer(to, operator_data) => {
+                    // transfer to a target address
+                    let transfer_params = TransferParams {
+                        to,
+                        amount: token_params.amount,
+                        operator_data,
+                    };
+                    let receipt = sdk::send::send(&Address::new_id(sdk::message::caller()), method_hash!("Transfer"), RawBytes::serialize(&transfer_params).unwrap(), TokenAmount::zero()).unwrap();
+                    // ignore failures at this level and return the transfer call receipt so caller can decide what to do
+                    return_ipld(&receipt)
+                }
+                TestAction::Burn => {
+                    // burn the tokens
+                    let burn_params = BurnParams {
+                        amount: token_params.amount,
+                    };
+                    let receipt = sdk::send::send(&Address::new_id(sdk::message::caller()), method_hash!("Burn"), RawBytes::serialize(&burn_params).unwrap(), TokenAmount::zero()).unwrap();
+                    if !receipt.exit_code.is_success() {
```

Comment:
Not sure if this should explode or just return the result upstream (so failure to burn will keep the tokens) like Transfer does above

Reply:
I would say it should abort and so should transfer, but I see you're using this return value in tests, so ok to keep this behaviour if it's useful.